### ISSUE TITLE
Add Seeking With Slider For tvOS

### ISSUE THREAD

**Issue by esegebart** [criszz77/react-native-media-console#49]
Hello! 

I wanted to add a suggestion to be able to seek using the slider bar with tvOS and remote. I didn't see it in the latest releases. 
Could this be added in the future? 

*Also - can you tell me how to grab this project and how to PR? I would like to try and do the seekbar thing, not sure if I can figure it out, but I can try!

Thank you!

**Comment by criszz77** [comment]
Hello @esegebart 

🙈 I'm so sorry but I didn't really have time to work on any tvOS support even though I promised so 🥺

**Here are are the steps I do to contribute to the project:**
1. Fork the project
2. Run `yarn` to install the packages.
3. Make your changes.

**In order to test them:**
1. Run `yarn prepare` to build the code. Since it's built with TypeScript, plain JS projects also have to be able understand them.
2. Delete `node_modules` from the forked `react-native-media-console` project, else you will end up with two instances of React and React Native and your project will break.
3. Copy the path of the forked `react-native-media-console` project, and run `yarn [path to fork]` in your project to install it. https://technastic.com/copy-file-path-mac/
4. `yarn start --reset-cache` and refresh your app! (I don't think you need to rebuild your app since it only contains JS code.)


Please let me know if you stumble on anything so I can improve those instructions!

Thank you very much and once again I am so sorry!

**Comment by poojan010** [comment]
Hey @criszz77 can you add this contributing info in Readme file

**Comment by esegebart** [comment]
@criszz77 It is totally okay! This will be a good chance for me to contribute! And these instructions will work for React Native tvOS? Do I need to do anything special for that or just add the tvOS package I am using to the package.json before I run yarn?

**Comment by criszz77** [comment]
@esegebart I updated my original post, I hope it's better now and didn't add to any confusion. Let me know how it goes! I will answer you ASAP.

**Comment by esegebart** [comment]
@criszz77 Thank you! That makes more sense now! Going to try some changes and see how it goes! So these are a couple things I was thinking about: 

1. The seeking forward by 15 seconds maybe be good for shorter videos, but not for longer videos. We have 12 hour broadcasts and their starting spot may be 8 hours into it and will be hard to get to with the 15 seconds or even 5 or 10 minutes, then they lose the granular control to get to their starting spot. So I want to try and do a progressive rewind / fast forward where if they navigate to the seekbar - and hold the right/left arrow it will rw/ff automatically start skipping forward at 2x, 4x  - or if they press it once it will do the 15 seconds. Haven't quite figures out exactly what I want it to look like but something maybe similar to other streaming apps like YouTube where the focus is automatically on the seekbar when the controls pops up - and leave the play button for play/pause. Youtube animates the play and pause button when you press the enter button so its there but just hidden until you hit the enter button to play/pause. 
2. Instead of the gradient background, have the full screen the dark gray see through color - on a lighter background the play and pause buttons are hard to see when they are in focus since they dim instead of highlight when they are focused. 
3. Invert the color focus of the play/pause and control buttons - when it goes into focus have it be opaque white and when out of focus do the half opacity. 

If these things sound ok to you, I will be happy to try working on them. :)



5.

**Comment by criszz77** [comment]
@esegebart 

1. That actually sounds good! Any tvOS support is welcomed, unfortunately I don't have the time to invest too much time in this, but I can give a helping hand if you have something in mind and actually manage to come up with a POC or something. Funny but I really don't use TV at all? So I also don't have a great idea of how user experience should be.

2. There was also this PR: https://github.com/criszz77/react-native-media-console/pull/43

It aimed to solve the issue of buttons on white background. Should be available with the newest version. Isn't this good enough? Do you have a better suggestion? 

3. Hmm, I noticed a bunch of work at `react-native-tvOS` regarding the focus. 

I'm pretty much not familiar with `tvOS` but we also need to make it work with or without `tvOS`, so importing `useTVEventHandler` in plain `react-native` projects might crash the app. I guess we can look into it to improve it once we have something working

**Comment by esegebart** [comment]
Ah yeah, that's what I am hoping I don't do is mess up support for anything
else. But it will be able to be tested.

I saw the overlay and have the latest version, but the overlay is like a
gradient color - where the dark is on the edges a lot but not the whole
screen. When I look at YouTube they have a very dark overlay, whereas this
isn't dark enough if there are white areas on the screen (which I had
encountered). I was looking at that component so that's super exciting it's
already there!

So I will try and get something working! And if it will work with both that
would be sweet, if it doesn't we can figure out where to go from there.

Silly question, but what is POC?

**Comment by criszz77** [comment]
Of course, don't worry. Give it a try, see what you're able to do and if you got something working I'll step in and help you finish.

Oh POC just stands for Proof Of Concept :D

**Comment by esegebart** [comment]
Okay, great! I will see what I can do. :)

**Comment by criszz77** [comment]
I'm sorry, I guess it's `yarn add [path to the esegebart/react-native-media-console]`

**Comment by esegebart** [comment]
Okay perfect! :) I will try that now.

**Comment by esegebart** [comment]
@criszz77 Ahh! I test my first changes! This is soooo exciting!

**Comment by criszz77** [comment]
@esegebart So happy to hear that! Let me know if I can help you with anything else! Hope you'll have fun

**Comment by esegebart** [comment]
Okay I have the small change for the opacity being darker on the overlay. How can I make a pull request to you?

**Comment by esegebart** [comment]
@criszz77 I have the idea sorted out kind of for this feature. The only thing I can't really figure out is how to get it to continuously skip at that seeking interval until the user either pushes the button again to go faster, or pushes play to stop where they are. And also still need to update the button to have 2x, 3x, 4x on the arrow button. 

Right now I have it skipping at the proper intervals.. for example, I press it once and it does the default rewindTime. I push it twice and it calculates a percentage of the video (1%) and subtracts from current time. 2nd press is 3% of the video. 3rd press is 4%. 4th press is 5%. 

Could you maybe check it out and tell me what I might need to do to get it to repeat the skipping interval until either the user presses play or hits the button again and tell me which way to go? 

**NOTE**: This only works for **rewind** time button currently, so you have move the slider to the middle and then start pushing the rewind button. I added in some console logs to show the time and the calculations too to see what it is doing *

**Comment by criszz77** [comment]
@esegebart Yeah, sure, I was planning to spend some time on this module this weekend and/or the next one. Have you got a fork prepared?

**Comment by esegebart** [comment]
I have it committed to my branch! Do I have to do anything special?

**Comment by criszz77** [comment]
@esegebart Open a pull request. Thank you so much for your time! I really appreciate it!

**Comment by esegebart** [comment]
@criszz77 Woohoo! :) Yes absolutely, this was a very interesting problem and I thank **you** for letting me be a part of it!

**Comment by esegebart** [comment]
@criszz77 By the way, the updated react native tvos has long presses for the arrow keys but I couldn't figure out how to get them to register. I updated to the new version that has them - but just another option to look at if you need it. I think the button presses will be just fine. :)

**Comment by criszz77** [comment]
Ohh, all right. I'll have to catch up with the tvOS environment, as I don't have any experience, but sure it will be fun! Anything else I should know before you go off on the weekend?

**Comment by esegebart** [comment]
@criszz77 All I'm trying to accomplish is for them to push the button one time to skip 15 sec, push it twice and it will keep skipping at 1% of video length, push a third time and it will skip 3%, push it 4 times and it skips 5%. When they push it a 5th time it just goes back to playing. I thought maybe having 2x, 3x and 4x on the buttons will be a nice indicator of what's happening. And if they push play it will stop the skipping. 

Here's the link to where I asked Doug about the long presses: 

https://github.com/react-native-tvos/react-native-tvos/discussions/409

And then here is where it was added: 

https://github.com/react-native-tvos/react-native-tvos/commit/1f1f338698f2c864c6d7c5b7169c1ed48e44f119

Pressing and holding right or left arrow key on the rewind/FF button for it to keep skipping would be great if it progressively speeds up after like 2sec of it being held it kicks into 2x after 5sec it kicks into 3x and 7sec kicks into 4x speed. That would be more familiar for user, but the button pressing is also good. 

I just don't think there is a way to focus the seekbar with tvos so using the rewind and fast forward buttons is good. But you are much better at React Native than I so there may be is a way but I couldn't solve it 😎

**Comment by esegebart** [comment]
@criszz77 Hi! Just wondering if you had a chance to look at this and can maybe give me some direction to continue working on it, or if you had worked on it yourself? :)

**Comment by criszz77** [comment]
Hey @esegebart 

Is this linked to https://github.com/criszz77/react-native-media-console/pull/59 ?

I am so sorry but I was so busy lately 🙈 I was hoping I could take a look this weekend, but the odds are against me. 🥺

**Comment by esegebart** [comment]
@criszz77 Yes, linked to #59 - and okay, I understand. :) Was just asking, I will see if I can keep picking away at it. I just got stuck trying to figure out how to continuously repeat the same interval until user does something else. I will research some more.

**Comment by esegebart** [comment]
@LunatiqueCoder  It is hereeeeee - https://www.npmjs.com/package/react-native-tvos/v/0.69.8-0rc3 

I can't wait to try and see if it does what I expect it to! :)

**Comment by LunatiqueCoder** [comment]
Hey @esegebart 

I just tried it out and the problems seems to be the same :( I will try again, maybe it's an issue on my side.

**Comment by esegebart** [comment]
@LunatiqueCoder Oh no, that is a bummer to hear.

**Comment by esegebart** [comment]
@LunatiqueCoder  Did you find out if it is still on the react native tvos side?

**Comment by LunatiqueCoder** [comment]
Hey @esegebart, I'm sorry I'm moving so hard but I also got involved with other projects 🙈

I'll try again this weekend hopefully. :) 🙏

**Comment by esegebart** [comment]
@LunatiqueCoder Okay! I was just checking in with you, I appreciate all of your help with this. I know you are a busy person!! :)

**Comment by esegebart** [comment]
@LunatiqueCoder I saw this comment a few days ago 

https://github.com/react-native-tvos/react-native-tvos/discussions/409#discussioncomment-5552606

Saying it was continuing to spawn the same event and didn't know when the key was being released. 

And Doug said this in the comment below him: 
you may want to turn on ReactFeatureFlags.enableKeyDownEvents. This will send a separate key up and key down event for each normal button press, and should send a key up at the end of the long press.

could this be what we needed??

**Comment by LunatiqueCoder** [comment]
Hey @esegebart 

That could be something :) 

I'm finally finishing with one of my projects so I will have more time starting this week. I really hope to release a new version this weekend if everything goes well. 🙏

**Comment by esegebart** [comment]
@LunatiqueCoder 

So exciting! I just wanted to try and stay involved and keep track of what was happening with the tvOS repo. :) Congrats on your project! I look forward to seeing if this helps for the slider! ❤️

**Comment by LunatiqueCoder** [comment]
@esegebart I had a look and indeed, it gets better, but it's still not perfect. We still need to handle `onPressOut` somehow: 

Source: https://reactnative.dev/docs/pressable#how-it-works
<img width="854" alt="image" src="https://user-images.githubusercontent.com/55203625/232723359-9e339929-9f6d-4885-8258-7ea1a24ae1b7.png">

However, with a few workarounds, I think you can achieve this without changing `react-native-media-console`, at least too much. Unfortunately I don't have the time to go that deep, but I'll try again whenever I have time.

**Comment by esegebart** [comment]
@LunatiqueCoder Then that's what I will have to do! You are so awesome 😎 And thank you so much for the diagram and resources. You have been so great to work with and I cannot thank you enough for teaching me open source! I'll be back to contribute more!

**Comment by esegebart** [comment]
@LunatiqueCoder So the longRight is now firing before the button is released, I am circling back to this and am going to finish it up. I just wanted to ask you again so I can refresh my memory, will the onPressOut handling code be inside react-native-media-console? 

I am thinking about longRight/longLeft - I will start seeking at 1x, after 2 second increase to 2x, after 2 seconds increase to 4x. And then I handle the onPressOut to resume playing. Is that where I would put this code is in PlayPause? Since I will be using the rewind and fast forward buttons? Thanks!

**Comment by LunatiqueCoder** [comment]
Hello @esegebart 

Yeah I can see they improved it. 🤔 I'll take a look sometime this weekend and come back with a response.

**Comment by esegebart** [comment]
@esegebart Ok! I am going to keep working on it in the meantime. I am using the 0.68.8-0 and tested the longRight and it fires promptly, and not after the button is released! Woohoo!

**Comment by esegebart** [comment]
@LunatiqueCoder OMGSH I JUST ABOUT HAVE IT. I am skipping forward when the right arrow is pressed, it updates the current time, and now I just need it to stop at the spot when I push play again bc it is currently going back to the original spot.

**Comment by esegebart** [comment]
@LunatiqueCoder I wrote the logic inside `react-native-media-console` `VideoPlayer `because I couldn't figure out how to get the remote events to pick up outside of it and interacting with the components was easier. I didn't have to change any of your code in the components to make it work except for what happens when `onPressRewind `is pressed- the logic is all separate to what happening in `VideoPlayer`. I might actually change that part so if its tvOS it calls my function otherwise do the logic you had and my logic can all live in the `useEffect `I am using for the `pressCount`. I can show you if you would like and you can diff it to see what is different. :)

**Comment by LunatiqueCoder** [comment]
@esegebart Yeah I would like to see that. I think that we can pass props to each button if you want by changing the project so you would just pass `onPressRewind` while your tvOS logic lives out of the `react-native-media-console`.

**Comment by esegebart** [comment]
Awesome, let me try and finish working out a doing the rewind also and I'll
send it to you.

**Comment by esegebart** [comment]
@LunatiqueCoder OKAY sooo I had another great idea! I saw you have a prop to disable seek bar controls. Do you think I could read the remote events on my video player end and add my own slider that pops open when the overlay pops open? Something like : 

```
import React, { useState, useEffect, useRef } from 'react';
import { View, TouchableOpacity } from 'react-native';
import Video from 'react-native-video';
import Slider from '@react-native-community/slider';
import { useTVEventHandler } from "react-native-tvos";

const VideoPlayer = () => {
  const [duration, setDuration] = useState(0.0);
  const [currentTime, setCurrentTime] = useState(0.0);
  const [isPaused, setIsPaused] = useState(false);
  const videoPlayer = useRef(null);

  const tvEventHandler = useTVEventHandler((evt) => {
    if (evt.eventType === 'right') {
      setCurrentTime(currentTime + 5);
      videoPlayer.current.seek(currentTime);
    }
    if (evt.eventType === 'left') {
      setCurrentTime(currentTime - 5);
      videoPlayer.current.seek(currentTime);
    }
  });

  const onLoad = (data) => {
    setDuration(data.duration);
  };

  const onProgress = (data) => {
    setCurrentTime(data.currentTime);
  };

  return (
    <View>
      <MediaPlayer
        ref={videoPlayer}
        source={{uri: 'http://commondatastorage.googleapis.com/gtv-videos-bucket/sample/BigBuckBunny.mp4'}}
        style={{height: 200, width: 300}}
        paused={isPaused}
        onLoad={onLoad}
        onProgress={onProgress}
        repeat={true}
        ...etc
      />

      <TouchableOpacity activeOpacity={1} onPress={() => setIsPaused(!isPaused)}>
        <Slider
          style={{width: 300, height: 40}}
          minimumValue={0}
          maximumValue={duration}
          value={currentTime}
          onValueChange={(value) => {
            setCurrentTime(value);
            videoPlayer.current.seek(value);
          }}
          minimumTrackTintColor="#FFFFFF"
          maximumTrackTintColor="#000000"
        />
      </TouchableOpacity>
    </View>
  );
};

export default VideoPlayer;
```
Idk just spitballing here as a plan B

**Comment by LunatiqueCoder** [comment]
I believe you shouldnbe able to do it, yes.

**Comment by esegebart** [comment]
@LunatiqueCoder Nice, okay. Well I am pretty happy with how things look with what I've been working on. I will get with you tomorrow and should have something ready then. :) Thanks Ovidiu!

**Comment by LunatiqueCoder** [comment]
Oh don’t worry, take your time. No pressure, this is open source 👌

**Comment by esegebart** [comment]
@LunatiqueCoder Okay, I pushed my changes to my `react-native-media-console` repo under the `seekbar-controls-feature` branch. It's not pretty or cleaned up yet, but it works, :D. It continuously skips forward and continuously skips backwards up to 4x speed, until play/pause is pressed to stop it. You will see the skip time increases each time its pressed. If its pushed a 5th time it will resume playing as well. Let me know what you think. Can the use effects and logic live outside the video player on my end, could this be of benefit to anyone, or can I use the controls like this out of my forked repo? I would like to add like 2x, 3x, 4x icons to show that the speed increases as they press. Anywho, I am pretty excited about this! :)

**Comment by esegebart** [comment]
@LunatiqueCoder I forgot to add white to the icons in the local repo! I just pushed those changes, but there's like 13 commits and I only want the last commit I did in the Pull Request. Is that possible?

**Comment by LunatiqueCoder** [comment]
I believe if you are familiar with rebasing, yes, that should be possible. But it can be tricky if it's the first time. I can do it for you, but here are some relevant links: 

https://www.w3docs.com/snippets/git/how-to-combine-multiple-commits-into-one-with-3-steps.html
https://stackoverflow.com/questions/45548100/how-to-start-typing-when-entering-vim-during-git-rebase

**Comment by esegebart** [comment]
@LunatiqueCoder I think I might have been looking at an old one from February ! Lol.

**Comment by LunatiqueCoder** [comment]
@esegebart I believe if you are familiar with rebasing, yes, that should be possible. But it can be tricky if it's the first time. I can do it for you, but here are some relevant links: 

https://www.w3docs.com/snippets/git/how-to-combine-multiple-commits-into-one-with-3-steps.html
https://stackoverflow.com/questions/45548100/how-to-start-typing-when-entering-vim-during-git-rebase

**Comment by esegebart** [comment]
@LunatiqueCoder Woohoo, I was also seeing the commits in reverse order so I thought I was missing my commits. Nope! Just me, looking at it wrong 🤦 😆 I made the pull request so phew! What a good day :)
